src/onnx/parse_matmul.cpp

```diff
@@ -43,66 +43,127 @@ struct parse_matmul : op_parser<parse_matmul>
+            }
 
-        instruction_ref bl0 = l0;
-        instruction_ref bl1 = l1;
-        if(not std::equal(
-               l0_lens.rbegin() + 2, l0_lens.rend(), l1_lens.rbegin() + 2, l1_lens.rend()))
-        {
-            auto l0_it = l0_lens.begin() + l0_lens.size() - 2;
-            std::vector<std::size_t> l0_broadcasted_lens(l0_lens.begin(), l0_it);
-            auto l1_it = l1_lens.begin() + l1_lens.size() - 2;
-            std::vector<std::size_t> l1_broadcasted_lens(l1_lens.begin(), l1_it);
-            auto output_lens = compute_broadcasted_lens(l0_broadcasted_lens, l1_broadcasted_lens);
-            l0_broadcasted_lens = output_lens;
-            l0_broadcasted_lens.insert(l0_broadcasted_lens.end(), l0_it, l0_lens.end());
-            l1_broadcasted_lens = output_lens;
-            l1_broadcasted_lens.insert(l1_broadcasted_lens.end(), l1_it, l1_lens.end());
-            if(l0_lens != l0_broadcasted_lens)
+            // TODO: handling this case requires a new multibroadcast mode
+            if(not std::equal(
+                   s0_dds.rbegin() + 2, s0_dds.rend(), s1_dds.rbegin() + 2, s1_dds.rend()))
+            {
+                MIGRAPHX_THROW("PARSE_MATMUL: dynamic shape broadcasting not supported");
+            }
+
+            instruction_ref dot_res = info.add_instruction(make_op(opd.op_name), a0, a1);
+
+            // squeeze the appended or prepended dimensions
+            int64_t num_axis = static_cast<int64_t>(dot_res->get_shape().ndim());
+            if(is_a_prepended)
             {
-                bl0 = info.add_instruction(
-                    make_op("multibroadcast", {{"out_lens", l0_broadcasted_lens}}), l0);
+                dot_res =
+                    info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 2}}}), dot_res);
+                --num_axis;
             }
-            if(l1_lens != l1_broadcasted_lens)
+            if(is_b_appended)
             {
-                bl1 = info.add_instruction(
-                    make_op("multibroadcast", {{"out_lens", l1_broadcasted_lens}}), l1);
+                dot_res =
+                    info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 1}}}), dot_res);
             }
+
+            return dot_res;
         }
-        instruction_ref dot_res = info.add_instruction(make_op(opd.op_name), bl0, bl1);
-        int64_t num_axis        = static_cast<int64_t>(dot_res->get_shape().lens().size());
-        if(is_a_prepended)
-        {
-            dot_res = info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 2}}}), dot_res);
-            --num_axis;
-        }
-        if(is_b_appended)
+        else
         {
-            dot_res = info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 1}}}), dot_res);
-        }
+            auto s0_lens = s0.lens();
+            auto s1_lens = s1.lens();
+
+            // args[0] is a vector, prepend 1 to the shape
```

Comment:
We should only prepend 1 when `s0.ndim() != s1.ndim()`.

Reply:
I think it's trying to do the same thing as numpy.matmul. The `dot` operator also only works for tensors with `ndim >= 2`.